modules/etc/image_etc.cpp

```diff
@@ -128,8 +128,9 @@ static void _compress_etc(Image *p_img, float p_lossy_quality, bool force_etc1_f
 
 	PoolVector<uint8_t>::Read r = img->get_data().read();
 
-	int target_size = Image::get_image_data_size(imgw, imgh, etc_format, p_img->has_mipmaps() ? -1 : 0);
 	int mmc = p_img->has_mipmaps() ? Image::get_image_required_mipmaps(imgw, imgh, etc_format) : 0;
```

Comment:
Ah wait, this should probably be replaced with
```cpp
int mmc = (p_img->has_mipmaps() ? Image::get_image_required_mipmaps(imgw, imgh, etc_format) : 0) + 1;
```

Reply:
Does `mmc` stand for "mipmaps count"? If so I guess it should just be this, no?
```cpp
int mmc = p_img->has_mipmaps() ? Image::get_image_required_mipmaps(imgw, imgh, etc_format) : 1;
```

Reply:
No there was little bit of inconsistency about the mipmap count, which is the source of confusion. Check it but if memory serves, get_image_required_mipmaps probably returns the number of mipmaps without counting the image with original size.

Reply:
If only the cpp code had some bit of documentation

Reply:
like at least what does this non-trivial function do

Reply:
BTW I wanted to check this with a recent build but my console is being flooded with

```
WARNING: not found: display/window/size/width
WARNING: not found: display/window/size/height
```

Is it just me or is this a known issue?

Reply:
Likely a regression from d09160a8b67fdc60e8108962c4e9bd4c0bc7f13e

Reply:
on second look would calling `int mmc = p_img->get_mipmap_count() + 1;` be ok here?

Reply:
Just checked, get_image_required_mipmaps does count the original size image as well, so it should indeed be
```cpp
int mmc = p_img->has_mipmaps() ? Image::get_image_required_mipmaps(imgw, imgh, etc_format) : 1;
```

or simply

```cpp
int mmc = Image::get_image_required_mipmaps(imgw, imgh, etc_format);
```

(given the work etc needs to do, that "optimization" isn't worth it anyway)

Reply:
@MillionOstrich there shouldn't be a +1 because get_mipmap_count already counts the original image.
Also, the required # of mipmaps depends on the format, so the # of mipmaps of the original and etc may not match

Reply:
5cb1d06 fixed the "display/window/size/width" warning BTW

Reply:
@tagcup does that mean we want to remove the +1 from the for loop condition too?
`for (int i = 0; i < mmc + 1; i++) {`

Reply:
looks that way
btw can you change the loop variable to j in the inner loop?

Reply:
can you also check `modules/squish/image_compress_squish.cpp`? the code there should be more or less the same there so if there is a problem here, squish code would also need to be fixed

Reply:
I had a search around the codebase, `image_compress_squish` is using <= in the for loop for both compress and decompress. 

I also had a search for get_mipmap_count() as that calls the same method underneath -
 everything seems to be using +1:
```core\image.cpp(309): ERR_FAIL_INDEX_V(p_mipmap, get_mipmap_count() + 1, -1);
drivers\gles3\rasterizer_storage_gles3.cpp(772): int mipmaps = ((texture->flags & VS::TEXTURE_FLAG_MIPMAPS) && img->has_mipmaps()) ? img->get_mipmap_count() + 1 : 1;
editor\import\resource_importer_texture.cpp(245): int mmc = image->get_mipmap_count() + 1;
editor\import\resource_importer_texture.cpp(274): int mmc = image->get_mipmap_count() + 1;
modules\pvr\texture_loader_pvr.cpp(214): for (int i = 0; i <= new_img->get_mipmap_count(); i++) {
modules\squish\image_compress_squish.cpp(49): int mm_count = p_image->get_mipmap_count();
//^ uses <= in the loop later on
```

Results for `get_image_required_mipmaps`:
```modules\etc\image_etc.cpp(132): int mmc = p_img->has_mipmaps() ? Image::get_image_required_mipmaps(imgw, imgh, etc_format) : 1;
modules\squish\image_compress_squish.cpp(151): int mm_count = p_image->has_mipmaps() ? Image::get_image_required_mipmaps(w, h, target_format) : 0;
scene\resources\texture.cpp(591): int mipmaps = Image::get_image_required_mipmaps(tw, th, format);
```

The only case i can't see a +1 is in `texture.cpp:  StreamTexture::_load_data`, and that also has a >1 check too - but i don't understand what is happening in that method.

Reply:
If you can confirm that it is indeed the case, the only change that is necessary would be

```cpp
p_img->create(imgw, imgh, p_img->has_mipmaps(), etc_format, dst_data);
```

Reply:
or if you'd like to wait, I can give a proper check during weekend